chalk-solve/src/clauses/syntactic_eq.rs

```diff
@@ -0,0 +1,165 @@
+use std::{iter, mem::replace};
+
+use chalk_engine::fallible::Fallible;
+use chalk_ir::{
+    cast::Cast,
+    fold::{shift::Shift, Fold, Folder, SuperFold},
+    interner::Interner,
+    AliasEq, AliasTy, Binders, BoundVar, DebruijnIndex, Goal, GoalData, Goals, ParameterKind,
+    ParameterKinds, ProgramClause, ProgramClauseData, ProgramClauseImplication, QuantifierKind, Ty,
+    TyData,
+};
+
+pub fn syn_eq_lower<I: Interner, T: Fold<I>>(interner: &I, clause: &T) -> <T as Fold<I>>::Result {
```

Comment:
Needs a comment. Also, I wonder if it's worth adding some *other* trait since you only really want to fold program clauses and goals here..? Nah, probably not, but we should add a comment about it.

Reply:
I think making it into `syn_lower_clause` and `syn_lower_goal` should be enough

Reply:
Yeah, two separate functions would be fine